### PR TITLE
refactor: extract commit sending into separate actor WPB-2091

### PR DESCRIPTION
**Pull request by David-Henner**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-2091" title="WPB-2091" target="_blank">StoryWPB-2091</a>  [iOS] put Certificate in KeyPackages and LeafNodes
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

This PR contains a refactoring of the logic for sending commit bundles

### Changes

- Introduced `CommitSender` to send commit bundles
- Created `CommitError` and `ExternalCommitError` to represent errors and recovery strategies
- Added an async method to `SafeCoreCrypto` in preparation for core crypto becoming async
- Updated `MLSActionsExecutor` to use `CommitSender`
- Updated `MLSService` to use new error types

### Testing

- Updated `MLSServiceTests`
- Updated `MLSActionsExecutorTests`
- Added tests for `CommitSender`

### CODE REVIEW

File: wire-ios-data-model/WireDataModel.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -288,6 +288,11 @@
 		6391A7FA2A6FD6FC00832665 /* MappingModel_2.106-2.107.xcmappingmodel in Sources */ = {isa = PBXBuildFile; fileRef = 6391A7F92A6FD6FC00832665 /* MappingModel_2.106-2.107.xcmappingmodel */; };
 		6391A7FD2A6FD7D100832665 /* DatabaseMigrationTests+UserClientUniqueness.swift in Sources */ = {isa = PBXBuildFile; fileRef = 6391A7FB2A6FD7C900832665 /* DatabaseMigrationTests+UserClientUniqueness.swift */; };
 		6391A7FF2A6FDB9100832665 /* store2-107-0.wiredatabase in Resources */ = {isa = PBXBuildFile; fileRef = 6391A7FE2A6FDB9100832665 /* store2-107-0.wiredatabase */; };
+		639971AE2B2732E6009DD5CF /* CommitSender.swift in Sources */ = {isa = PBXBuildFile; fileRef = 639971AD2B2732E6009DD5CF /* CommitSender.swift */; };
+		639971B02B27346B009DD5CF /* CommitError.swift in Sources */ = {isa = PBXBuildFile; fileRef = 639971AF2B27346B009DD5CF /* CommitError.swift */; };
+		639971B22B2734C0009DD5CF /* ExternalCommitError.swift in Sources */ = {isa = PBXBuildFile; fileRef = 639971B12B2734C0009DD5CF /* ExternalCommitError.swift */; };
+		639971BF2B28BB63009DD5CF /* config.yml in Resources */ = {isa = PBXBuildFile; fileRef = 639971BE2B28BB63009DD5CF /* config.yml */; };
```
Comment (caldrian) [inline wire-ios-data-model/WireDataModel.xcodeproj/project.pbxproj]: In my opinion the `config.yml` file should not be added to the project, unless you find it very convenient to be able to edit it in Xcode. However, we probably don't edit these files often.

File: wire-ios-data-model/Source/MLS/CommitSender.swift [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+
+            WireLogger.mls.info("merging commit for group (\(groupID.safeForLoggingDescription))")
+            try await mergeCommit(in: groupID)
+
+            return events
+
+        } catch let error as SendCommitBundleAction.Failure {
+            WireLogger.mls.warn("failed to send commit bundle: \(String(describing: error))")
+
+            let recoveryStrategy = CommitError.Recovery(from: error)
```
Comment (caldrian) [inline wire-ios-data-model/Source/MLS/CommitSender.swift]: Should `CommitError.Recovery` also be named `CommitError.RecoveryStrategy` instead?

File: wire-ios-data-model/Tests/MLS/CommitSenderTests.swift [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+
+        // Then
+        await assertItThrows(error: ExternalCommitError.failedToSendCommit(recovery: recovery)) {
+            // When
+            _ = try await sut.sendExternalCommitBundle(commitBundle, for: groupID)
+        }
+
+        if shouldClearPendingGroup {
+            // It clears pending commit
+            XCTAssertEqual(mockClearPendingGroupInvocations.count, 1)
```
Comment (caldrian) [inline wire-ios-data-model/Tests/MLS/CommitSenderTests.swift]: The `file` and `line` arguments are not used.

**Review by caldrian — APPROVED** [review]

**Comment by netbe** [comment]
@David-Henner could you give a bit more info on why this refactoring is needed ?

**Comment by David-Henner** [comment]
> @David-Henner could you give a bit more info on why this refactoring is needed ?

Sure. We need the ability to send commits for E2EI. By doing this refactoring we make it possible to send commits without having the `MLSService` or `MLSActionExecutor` as dependencies in the classes related to E2EI

**Review by netbe — APPROVED** [review]

File: wire-ios-data-model/Tests/MLS/CommitSenderTests.swift (hunk 3 above)
Comment (David-Henner) [reply wire-ios-data-model/Tests/MLS/CommitSenderTests.swift]: good catch! fixed here https://github.com/wireapp/wire-ios/pull/787/commits/b68cde7421b9d6c96ddab4468e3e05b899778eaf

File: wire-ios-data-model/Source/MLS/CommitSender.swift (hunk 2 above)
Comment (David-Henner) [reply wire-ios-data-model/Source/MLS/CommitSender.swift]: done https://github.com/wireapp/wire-ios/pull/787/commits/b68cde7421b9d6c96ddab4468e3e05b899778eaf

**Pull request merged**